Bug Report: deployment fails with NPE when MPC value is only available profiled

### Brief Summary

We use a MicroProfile Config lookup in a constructur. The setting is available from `microprofile-config.properties`, but only as a profiled setting: `%ct.dataverse.siteUrl`.

The deployment of Dataverse fails with an NPE because the resolver does not find a non-profiled value.

### Expected Outcome

Just use the profiled value if no other value is present.

### Current Outcome

```
Caused by: java.lang.NullPointerException: Cannot invoke "fish.payara.nucleus.microprofile.config.spi.ConfigValueImpl.getSourceOrdinal()" because "resultWithoutProfile" is null
 	at fish.payara.nucleus.microprofile.config.spi.ConfigExpressionResolver.resolve(ConfigExpressionResolver.java:112)
 	at fish.payara.nucleus.microprofile.config.spi.ConfigExpressionResolver.resolveExpression(ConfigExpressionResolver.java:227)
 	at fish.payara.nucleus.microprofile.config.spi.ConfigExpressionResolver.resolveExpression(ConfigExpressionResolver.java:153)
 	at fish.payara.nucleus.microprofile.config.spi.ConfigExpressionResolver.getValue(ConfigExpressionResolver.java:135)
 	at fish.payara.nucleus.microprofile.config.spi.ConfigExpressionResolver.resolve(ConfigExpressionResolver.java:108)
 	at fish.payara.nucleus.microprofile.config.spi.ConfigExpressionResolver.resolve(ConfigExpressionResolver.java:91)
 	at fish.payara.nucleus.microprofile.config.spi.PayaraConfig.searchConfigSources(PayaraConfig.java:352)
	at fish.payara.nucleus.microprofile.config.spi.PayaraConfig.getConfigValue(PayaraConfig.java:251)
 	at fish.payara.nucleus.microprofile.config.spi.PayaraConfig.getConfigValue(PayaraConfig.java:229)
 	at fish.payara.nucleus.microprofile.config.spi.PayaraConfig.getValue(PayaraConfig.java:224)
 	at fish.payara.nucleus.microprofile.config.spi.PayaraConfig.getValueInternal(PayaraConfig.java:200)
 	at fish.payara.nucleus.microprofile.config.spi.PayaraConfig.getOptionalValue(PayaraConfig.java:173)
```

### Reproducer

I will at least provide a PR with a unit test demonstrating the fail. I will try to provide a fix as well.

### Operating System

Linux

### JDK Version

Temurin 17.0.9

### Payara Distribution

Payara Server Full Profile

--- a/nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/spi/ConfigExpressionResolver.java
+++ b/nucleus/payara-modules/nucleus-microprofile/config-service/src/main/java/fish/payara/nucleus/microprofile/config/spi/ConfigExpressionResolver.java
@@ -109,8 +109,12 @@ final class ConfigExpressionResolver {
         
         if(profile != null && result != null) {
             ConfigValueImpl resultWithoutProfile = getValue(resolveExpression(propertyName));
-            result = resultWithoutProfile.getSourceOrdinal() == result.getSourceOrdinal() ? result :
-                    (resultWithoutProfile.getSourceOrdinal() > result.getSourceOrdinal()) ? resultWithoutProfile : result;
+            // Note: In case there is a non-profiled value from a source with the same ordinal value, it will be ignored.
+            //       All spec versions including v3.1 do not include a definition for this edge case -
+            //       all sources are supposed to have a unique ordinal value.
+            if (resultWithoutProfile != null && resultWithoutProfile.getSourceOrdinal() > result.getSourceOrdinal()) {
+                result = resultWithoutProfile;
+            }
         } 
         
 

--- a/nucleus/payara-modules/nucleus-microprofile/config-service/src/test/java/fish/payara/nucleus/microprofile/config/spi/ConfigExpressionResolverTest.java
+++ b/nucleus/payara-modules/nucleus-microprofile/config-service/src/test/java/fish/payara/nucleus/microprofile/config/spi/ConfigExpressionResolverTest.java
@@ -44,7 +44,9 @@ import org.eclipse.microprofile.config.spi.ConfigSource;
 import org.junit.Before;
 import org.junit.Test;
 
+import java.util.Collections;
 import java.util.HashMap;
+import java.util.Set;
 
 import static fish.payara.nucleus.microprofile.config.spi.ConfigTestUtils.createSource;
 import static java.util.Collections.singleton;
@@ -55,7 +57,8 @@ import static org.junit.Assert.assertNull;
 public class ConfigExpressionResolverTest {
     
     private final ConfigSource source = createSource("S1", 100, new HashMap<>());
-    private final ConfigExpressionResolver resolver = new ConfigExpressionResolver(singleton(source), "test");
+    private final ConfigSource source2 = createSource("S2", 110, new HashMap<>());
+    private final ConfigExpressionResolver resolver = new ConfigExpressionResolver(Set.of(source, source2), "test");
 
     @Before
     public void configureConfigProperties() {
@@ -74,6 +77,10 @@ public class ConfigExpressionResolverTest {
         source.getProperties().put("default.key.reference", "${${not.existing:key}:not.found}");
         source.getProperties().put("%test.fish.payara.badger", "mushroom");
         source.getProperties().put("fish.payara.badger", "badger");
+        source.getProperties().put("%test.fish.payara.rod", "bites");
+        
+        source2.getProperties().put("fish.payara.rod", "nobites");
+        source2.getProperties().put("%test.fish.payara.profile-only", "gotcha");
     }
 
     @Test
@@ -183,6 +190,20 @@ public class ConfigExpressionResolverTest {
         ConfigValue result = resolver.resolve("fish.payara.badger");
         assertEquals("mushroom", result.getValue());
         assertEquals("mushroom", result.getRawValue());
+    }
+    
+    @Test
+    public void testProfilesOverrideProfiledValueFromSourceWithHigherOrdinal() {
+        ConfigValue result = resolver.resolve("fish.payara.rod");
+        assertEquals("nobites", result.getValue());
+        assertEquals("nobites", result.getRawValue());
+    }
+    
+    @Test
+    public void testProfilesDoNotFailToLookupWhenOnlyProfiledValueExists() {
+        ConfigValue result = resolver.resolve("fish.payara.profile-only");
+        assertEquals("gotcha", result.getValue());
+        assertEquals("gotcha", result.getRawValue());
     }
     
 }

--- a/nucleus/payara-modules/nucleus-microprofile/config-service/src/test/java/fish/payara/nucleus/microprofile/config/spi/ConfigExpressionResolverTest.java
+++ b/nucleus/payara-modules/nucleus-microprofile/config-service/src/test/java/fish/payara/nucleus/microprofile/config/spi/ConfigExpressionResolverTest.java
@@ -44,7 +44,6 @@ import org.eclipse.microprofile.config.spi.ConfigSource;
 import org.junit.Before;
 import org.junit.Test;
 
-import java.util.Collections;
 import java.util.HashMap;
 import java.util.Set;
 
@@ -57,8 +56,10 @@ import static org.junit.Assert.assertNull;
 public class ConfigExpressionResolverTest {
     
     private final ConfigSource source = createSource("S1", 100, new HashMap<>());
-    private final ConfigSource source2 = createSource("S2", 110, new HashMap<>());
-    private final ConfigExpressionResolver resolver = new ConfigExpressionResolver(Set.of(source, source2), "test");
+    private final ConfigSource sourceHigherOrdinal = createSource("SHO", 110, new HashMap<>());
+    private final ConfigSource sourceEqualOrdinal = createSource("SEO", 100, new HashMap<>());
+    private final ConfigSource sourceLowerOrdinal = createSource("SEO", 90, new HashMap<>());
+    private final ConfigExpressionResolver resolver = new ConfigExpressionResolver(Set.of(source, sourceHigherOrdinal, sourceEqualOrdinal, sourceLowerOrdinal), "test");
 
     @Before
     public void configureConfigProperties() {
@@ -79,8 +80,11 @@ public class ConfigExpressionResolverTest {
         source.getProperties().put("fish.payara.badger", "badger");
         source.getProperties().put("%test.fish.payara.rod", "bites");
         
-        source2.getProperties().put("fish.payara.rod", "nobites");
-        source2.getProperties().put("%test.fish.payara.profile-only", "gotcha");
+        sourceHigherOrdinal.getProperties().put("fish.payara.rod", "nobites");
+        sourceHigherOrdinal.getProperties().put("%test.fish.payara.profile-only", "gotcha");
+        
+        sourceEqualOrdinal.getProperties().put("fish.payara.badger", "i-shall-be-ignored");
+        sourceLowerOrdinal.getProperties().put("fish.payara.badger", "i-shall-be-ignored");
     }
 
     @Test
@@ -187,6 +191,8 @@ public class ConfigExpressionResolverTest {
     
     @Test
     public void testProfiles() {
+        // This test case does not only test if the profiled value is used from the same source, but also if the
+        // defined equal and lower ordinal sources are ignored
         ConfigValue result = resolver.resolve("fish.payara.badger");
         assertEquals("mushroom", result.getValue());
         assertEquals("mushroom", result.getRawValue());

